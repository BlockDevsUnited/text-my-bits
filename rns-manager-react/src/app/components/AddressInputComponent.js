import React, { useState } from 'react';
import propTypes from 'prop-types';
import { Button } from 'react-bootstrap';
import { toChecksumAddress } from 'rskjs-util';

import { validateAddress } from '../validations';
import { ChecksumErrorContainer } from '../containers';
import UserErrorComponent from './UserErrorComponent';
import UserSuccessComponent from './UserSuccessComponent';
import UserWaitingComponent from './UserWaitingComponent';

import edit from '../../assets/img/edit.svg';
import editActive from '../../assets/img/edit-active.svg';
import closeBlue from '../../assets/img/close-blue.svg';

const AddressInputComponent = ({
  allowDelete,
  label,
  labelDisplay,
  labelIcon,
  value,
  isWaiting,
  isError,
  handleErrorClose,
  handleSuccessClose,
  handleSubmit,
  handleDelete,
  isSuccess,
  strings,
  successTx,
  reset,
  validationChainId,
  validation,
}) => {
  const [isEditing, setIsEditing] = useState(false);
  const [isDeleting, setIsDeleting] = useState(false);
  const [isChecksumError, setIsChecksumError] = useState(false);
  const [isLocalError, setIsLocalError] = useState(false);
  const [editText, setEditText] = useState('');

  const handleEditClick = () => {
    setIsEditing(!isEditing);
    setIsDeleting(false);
  };

  const handleDeleteClick = () => {
    setIsDeleting(!isDeleting);
    setIsEditing(false);
  };

  const confirmDelete = () => {
    handleDelete();
  };

  const handleErrorClick = () => {
    setIsLocalError('');
    handleErrorClose();
  };

  const handleSubmitClick = () => {
    if (!validation) {
      return handleSubmit(editText);
    }

    if (editText.toLowerCase() === value.toLowerCase()) {
      return setIsLocalError('Value is the same.');
    }

    switch (validateAddress(editText, validationChainId)) {
      case 'Invalid address':
        return setIsLocalError('Invalid address');
      case 'Invalid checksum':
        return setIsChecksumError(true);
      default:
    }
    setIsLocalError(false);
    setIsChecksumError(false);
    return handleSubmit(editText);
  };

  const handleTextChange = (evt) => {
    setEditText(evt.target.value);
  };

  const handleChecksumClick = () => {
    setEditText(editText.toLowerCase());
    setIsChecksumError(false);
    return handleSubmit(editText);
  };

  const handleCancelClick = () => {
    setIsEditing(false);
    setIsChecksumError(false);
  };

  if (reset && (isEditing || isDeleting)) {
    setIsEditing(false);
    setIsDeleting(false);
    setEditText('');
  }

  return (
    <div className="row addressInput">
      <div className="row view">
        <div className="col-md-3 label">
          {labelIcon && <img src={labelIcon} alt={labelDisplay || label} />}
          {labelDisplay || label}
        </div>
        <div className={`${allowDelete ? 'col-md-7' : 'col-md-8'} value`}>
          {strings.value_prefix
            && <span className="value-prefix">{`${strings.value_prefix}: `}</span>
          }
          {validation ? toChecksumAddress(value, validationChainId) : value}
        </div>
        <div className={`${allowDelete ? 'col-md-2' : 'col-md-1'} options`}>
          <button
            type="button"
            onClick={handleEditClick}
            className="edit"
            disabled={isWaiting}
          >
            <img src={(!isEditing ? edit : editActive)} alt={strings.edit} />
          </button>
          {allowDelete
            && (
            <button
              type="button"
              onClick={handleDeleteClick}
              className="delete"
              disabled={isWaiting}
            >
              <img src={closeBlue} alt={strings.delete} />
            </button>
            )
          }
        </div>
      </div>
      {isEditing
        && (
        <div className="row edit">
          <div className="col-md-2 editLabel">
            {strings.edit_propmt}
          </div>
          <div className="col-md-6">
            <input
              type="text"
              placeholder={strings.edit_placeholder}
              value={editText}
              onChange={handleTextChange}
              disabled={isWaiting}
            />
          </div>
          <div className="col-md-4 buttons">
            <Button
              className="cancel"
              onClick={() => handleCancelClick()}
              disabled={isWaiting}
            >
              {strings.cancel}
            </Button>
            <Button
              className="submit"
              disabled={isWaiting}
              onClick={handleSubmitClick}
            >
              {strings.submit}
            </Button>
          </div>
        </div>
        )
      }
      {isChecksumError
        && (
          <ChecksumErrorContainer
            show={isChecksumError}
            inputValue={editText}
            handleClick={() => handleChecksumClick()}
          />
        )
      }
      {isDeleting
        && (
          <div className="delete">
            <p>{strings.delete_confirm_text}</p>
            <p>
              <Button
                className="cancel"
                onClick={() => { setIsDeleting(false); setIsLocalError(''); }}
                disabled={isWaiting}
              >
                {strings.cancel}
              </Button>
              <Button
                className="submit"
                disabled={isWaiting}
                onClick={confirmDelete}
              >
                {strings.delete}
              </Button>
            </p>
          </div>
        )
      }

      <UserWaitingComponent
        message={strings.waiting}
        visible={isWaiting}
      />

      <UserErrorComponent
        title={strings.error_title}
        message={isLocalError || strings.error_message}
        handleCloseClick={() => handleErrorClick()}
        visible={isError || isLocalError}
      />

      <UserSuccessComponent
        title={strings.success_title}
        message={strings.success_message}
        handleCloseClick={handleSuccessClose}
        address={successTx}
        visible={isSuccess}
      />
    </div>
  );
};

AddressInputComponent.defaultProps = {
  allowDelete: true,
  isError: false,
  isWaiting: false,
  isSuccess: false,
  reset: false,
  successTx: '',
  validation: true,
  validationChainId: null,
  strings: {
    cancel: 'Cancel',
    delete: 'Delete',
    delete_confirm_text: 'Are you sure you want to delete?',
    edit: 'Edit',
    edit_placeholder: 'enter address',
    edit_propmt: 'Change ownership',
    error_title: 'Error Title',
    error_message: 'Error Message',
    submit: 'Submit',
    success_title: 'Success Title',
    success_message: 'Success Message',
    value_prefix: 'Owner',
    waiting: 'Waiting text',
  },
  handleDelete: () => {},
  handleErrorClose: () => {},
  handleSuccessClose: () => {},
  labelDisplay: null,
  labelIcon: null,
};

AddressInputComponent.propTypes = {
  allowDelete: propTypes.bool,
  label: propTypes.string.isRequired,
  labelDisplay: propTypes.string,
  labelIcon: propTypes.string,
  isError: propTypes.bool,
  isWaiting: propTypes.bool,
  isSuccess: propTypes.bool,
  reset: propTypes.bool,
  validation: propTypes.bool,
  successTx: propTypes.string,
  handleErrorClose: propTypes.func,
  handleSuccessClose: propTypes.func,
  handleSubmit: propTypes.func.isRequired,
  handleDelete: propTypes.func,
  value: propTypes.string.isRequired,
  validationChainId: propTypes.string,
  strings: propTypes.shape(),
};

export default AddressInputComponent;
