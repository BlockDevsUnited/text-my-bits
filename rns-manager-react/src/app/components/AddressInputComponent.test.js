import React from 'react';
import { shallow } from 'enzyme';

import AddressInputComponent from './AddressInputComponent';

describe('AddressInputComponent', () => {
  const initProps = {
    allowDelete: true,
    label: 'Label',
    value: 'Value',
    handleErrorClose: jest.fn(),
    handleSuccessClose: jest.fn(),
    handleSubmit: jest.fn(),
    handleDelete: jest.fn(),
    validation: false,
    strings: {
      cancel: 'cancel string',
      delete: 'delete string',
      delete_confirm_text: 'delete confirm text',
      edit: 'edit text',
      edit_placeholder: 'placeholder text',
      edit_propmt: 'edit prompt text',
      error_title: 'error title text',
      error_message: 'error message text',
      submit: 'submit text',
      success_title: 'success title text',
      success_message: 'success message text',
      value_prefix: 'value prefix',
      waiting: 'waiting text string',
    },
  };

  const checksumInitialProps = {
    ...initProps,
    validation: true,
    strings: {
      ...initProps.strings,
      value_prefix: '',
    },
  };

  it('renders without crashing', () => {
    const component = shallow(<AddressInputComponent {...initProps} />);
    expect(component).toMatchSnapshot();
  });

  it('shows confirmation window when delete is clicked', () => {
    const component = shallow(<AddressInputComponent {...initProps} />);
    component.find('button.delete').simulate('click');
    expect(component.find('div.delete').length).toBe(1);
    expect(component).toMatchSnapshot();
  });

  it('delete is not shown when allowDelete is false', () => {
    const localProps = {
      ...initProps,
      allowDelete: false,
    };
    const component = shallow(<AddressInputComponent {...localProps} />);
    expect(component.find('button.delete').length).toBe(0);
  });

  it('start, edit, and delete allow custom text ', () => {
    const localProps = {
      ...initProps,
      strings: {
        cancel: 'custom cancel string',
        delete: 'custom delete string',
        delete_confirm_text: 'custom delete confirm text',
        edit: 'custom edit text',
        edit_placeholder: 'custom placeholder text',
        edit_propmt: 'custom edit prompt text',
        submit: 'custom submit text',
        value_prefix: 'custom value prefix',
      },
    };
    const component = shallow(<AddressInputComponent {...localProps} />);
    expect(component.find('div.value').text()).toEqual('custom value prefix: Value');

    // edit screen
    component.find('button.edit').simulate('click');
    expect(component.find('div.editLabel').text()).toEqual('custom edit prompt text');

    // delete screen
    component.find('button.delete').simulate('click');
    expect(component.find('div.delete').find('p').first().text())
      .toEqual('custom delete confirm text');
  });

  it('dispalys an icon correctly', () => {
    const localProps = {
      ...initProps,
      label: 'rsk',
      labelIcon: '/assets/icons/icon_rsk.png',
    };
    const component = shallow(<AddressInputComponent {...localProps} />);
    const image = component.find('div.label').find('img');
    expect(image.props().src).toEqual('/assets/icons/icon_rsk.png');
    expect(image.props().alt).toEqual('rsk');
  });

  it('displays correct checksum for ethereum', () => {
    const ethereumChecksum = '0xEe3D5f22Ea0FF393AeEf5Cf88a81E7d44979633B';

    const localProps = {
      ...checksumInitialProps,
      value: '0xee3d5f22ea0ff393aeef5cf88a81e7d44979633b',
    };

    const component = shallow(<AddressInputComponent {...localProps} />);
    expect(component.find('div.value').text()).toBe(ethereumChecksum);
  });

  it('displays correct checksum for RSK testnet', () => {
    const rskTestnetChecksum = '0xEE3D5f22Ea0Ff393aeeF5cf88a81E7D44979633B';

    const localProps = {
      ...checksumInitialProps,
      value: '0xee3d5f22ea0ff393aeef5cf88a81e7d44979633b',
      validationChainId: '31',
    };

    const component = shallow(<AddressInputComponent {...localProps} />);
    expect(component.find('div.value').text()).toBe(rskTestnetChecksum);
  });

  it('displays correct checksum for RSK mainnet', () => {
    const rskMainnetChecksum = '0x5215d879F378c902E6CC0CB9ace0240Ac7a863E7';

    const localProps = {
      ...checksumInitialProps,
      value: '0x5215d879f378c902e6cc0cb9ace0240ac7a863e7',
      validationChainId: '30',
    };

    const component = shallow(<AddressInputComponent {...localProps} />);
    expect(component.find('div.value').text()).toBe(rskMainnetChecksum);
  });
});
