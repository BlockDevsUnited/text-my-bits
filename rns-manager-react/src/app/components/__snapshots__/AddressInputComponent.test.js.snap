// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressInputComponent renders without crashing 1`] = `
<div
  className="row addressInput"
>
  <div
    className="row view"
  >
    <div
      className="col-md-3 label"
    >
      Label
    </div>
    <div
      className="col-md-7 value"
    >
      <span
        className="value-prefix"
      >
        value prefix: 
      </span>
      Value
    </div>
    <div
      className="col-md-2 options"
    >
      <button
        className="edit"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <img
          alt="edit text"
          src="mock-image"
        />
      </button>
      <button
        className="delete"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <img
          alt="delete string"
          src="mock-image"
        />
      </button>
    </div>
  </div>
  <UserWaitingComponent
    message="waiting text string"
    visible={false}
  />
  <Connect(UserErrorComponent)
    handleCloseClick={[Function]}
    message="error message text"
    title="error title text"
    visible={false}
  />
  <Connect(UserSuccessComponent)
    address=""
    handleCloseClick={[MockFunction]}
    message="success message text"
    title="success title text"
    visible={false}
  />
</div>
`;

exports[`AddressInputComponent shows confirmation window when delete is clicked 1`] = `
<div
  className="row addressInput"
>
  <div
    className="row view"
  >
    <div
      className="col-md-3 label"
    >
      Label
    </div>
    <div
      className="col-md-7 value"
    >
      <span
        className="value-prefix"
      >
        value prefix: 
      </span>
      Value
    </div>
    <div
      className="col-md-2 options"
    >
      <button
        className="edit"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <img
          alt="edit text"
          src="mock-image"
        />
      </button>
      <button
        className="delete"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <img
          alt="delete string"
          src="mock-image"
        />
      </button>
    </div>
  </div>
  <div
    className="delete"
  >
    <p>
      delete confirm text
    </p>
    <p>
      <Button
        active={false}
        className="cancel"
        disabled={false}
        onClick={[Function]}
        type="button"
        variant="primary"
      >
        cancel string
      </Button>
      <Button
        active={false}
        className="submit"
        disabled={false}
        onClick={[Function]}
        type="button"
        variant="primary"
      >
        delete string
      </Button>
    </p>
  </div>
  <UserWaitingComponent
    message="waiting text string"
    visible={false}
  />
  <Connect(UserErrorComponent)
    handleCloseClick={[Function]}
    message="error message text"
    title="error title text"
    visible={false}
  />
  <Connect(UserSuccessComponent)
    address=""
    handleCloseClick={[MockFunction]}
    message="success message text"
    title="success title text"
    visible={false}
  />
</div>
`;
